petab/sbml.py

```diff
@@ -98,3 +98,199 @@ def constant_species_to_parameters(sbml_model):
+            creating global parameters
+    """
+    for reaction in sbml_model.getListOfReactions():
+        law = reaction.getKineticLaw()
+        # copy first so we can delete in the following loop
+        local_parameters = [local_parameter for local_parameter
+                            in law.getListOfParameters()]
+        for lp in local_parameters:
+            if prepend_reaction_id:
+                parameter_id = f'{reaction.getId()}_{lp.getId()}'
+            else:
+                parameter_id = lp.getId()
+
+            # Create global
+            p = sbml_model.createParameter()
+            p.setId(parameter_id)
+            p.setName(lp.getName())
+            p.setConstant(lp.getConstant())
+            p.setValue(lp.getValue())
+            p.setUnits(lp.getUnits())
+
+            # removeParameter, not removeLocalParameter!
+            law.removeParameter(lp.getId())
+
+
+def add_global_parameter(sbml_model: libsbml.Model,
+                         parameter_id: str,
+                         parameter_name=None,
+                         constant=False,
+                         units='dimensionless',
+                         value=0.0):
+    """Add new global parameter to SBML model"""
+
+    if parameter_name is None:
+        parameter_name = parameter_id
+
+    p = sbml_model.createParameter()
+    p.setId(parameter_id)
+    p.setName(parameter_name)
+    p.setConstant(constant)
+    p.setValue(value)
+    p.setUnits(units)
+    return p
+
+
+def create_assigment_rule(sbml_model: libsbml.Model,
+                          assignee_id: str,
+                          formula:str,
+                          rule_id=None,
+                          rule_name=None):
+    """Create SBML AssignmentRule
+
+    Arguments:
+        sbml_model: Model to add output to
+        assignee_id: Target of assignment
+        formula: Formula string for model output
+        rule_id: SBML id for created rule
+        rule_name: SBML name for created rule
+    """
+    if rule_id is None:
+        rule_id = assignee_id
+
+    if rule_name is None:
+        rule_name = rule_id
+
+    rule = sbml_model.createAssignmentRule()
+    rule.setId(rule_id)
+    rule.setName(rule_name)
+    rule.setVariable(assignee_id)
+    rule.setFormula(formula)
+
+    return rule
+
+
+def add_model_output(sbml_model: libsbml.Model,
+                     observable_id: str,
+                     formula: str,
+                     observable_name: None):
+    """Add PEtab-style output to model
+
+    We expect that all formula parameters are added to the model elsewhere.
+
+    Arguments:
+        sbml_model: Model to add output to
+        formula: Formula string for model output
+        observable_id: ID without 'observable_' prefix
+        observable_name: Any observable name
+    """
+
+    if observable_name is None:
+        observable_name = observable_id
+
+    prefixed_id = f'observable_{observable_id}'
```

Comment:
check whether parameter / assignment rule exists in model already?

Reply:
I will leave that to the user. If libsbml will not complain, I will not complain. When modifying the model, it is generally a good idea to check the result afterwards, see respective functions added here.